imports/client/components/FancyEditor.tsx

```diff
@@ -0,0 +1,676 @@
+  editor.isInline = (element) => {
+    return element.type === 'mention' ? true : isInline(element);
+  };
+  editor.isVoid = (element) => {
+    return element.type === 'mention' ? true : isVoid(element);
+  };
+  editor.markableVoid = (element) => {
+    return element.type === 'mention' || markableVoid(element);
+  };
+  return editor;
+};
+
+// A normalization pass to merge all `message` elements into a single one.
+// Normal editing can't create more than one, but copy-pasting a break would
+// normally get transformed into a hard break, so we solve this by merging
+// sibling message nodes.
+const withSingleMessage = (editor: Editor) => {
+  const { normalizeNode } = editor;
+  editor.normalizeNode = (entry: NodeEntry) => {
+    const [node, _path] = entry;
+    if (Editor.isEditor(node)) {
+      if (node.children.length > 1) {
+        Transforms.mergeNodes(editor, { at: [1] });
+      }
+    }
+
+    normalizeNode(entry);
+  };
+  return editor;
+};
+/* eslint-enable no-param-reassign */
+
+const insertMention = (editor: Editor, userId: string) => {
+  const mention: MentionElement = {
+    type: 'mention',
+    userId,
+    children: [{ text: '' }],
+  };
+  Transforms.insertNodes(editor, mention);
+  Transforms.move(editor);
+};
+
+const MatchCandidateRow = styled.div<{ selected: boolean }>`
+  padding: 2px 3px;
+  border-radius: 3px;
+  height: 28px;
+  display: flex;
+  flex-direction: row;
+  align-items: center;
+  justify-content: flex-start;
+  cursor: pointer;
+  ${({ selected }) => css`
+    background: ${selected ? '#e0ecfc' : 'transparent'};
+  `}
+`;
+
+const MatchCandidateDisplayName = styled.div`
+  margin-left: 0.25rem;
+`;
+
+const MatchCandidate = ({
+  user, selected, onSelected,
+}: {
+  user: Meteor.User,
+  selected: boolean,
+  onSelected: (u: Meteor.User) => void,
+}) => {
+  const onClick = useCallback(() => {
+    onSelected(user);
+  }, [onSelected, user]);
+  return (
+    <MatchCandidateRow
+      key={user._id}
+      selected={selected}
+      onClick={onClick}
+    >
+      <Avatar size={24} {...user} />
+      <MatchCandidateDisplayName>{user.displayName}</MatchCandidateDisplayName>
+    </MatchCandidateRow>
+  );
+};
+
+type AugmentedUser = Meteor.User & {
+  foundDisplayName: boolean;
+  startsDisplayName: boolean;
+  foundEmail: boolean;
+  startsEmail: boolean;
+  foundGoogle: boolean;
+  startsGoogle: boolean;
+  foundDiscord: boolean;
+  startsDiscord: boolean;
+};
+
+function matchUsers(users: Meteor.User[], searchString: string): AugmentedUser[] {
+  // No point doing all this matching work if there's no search string to match against.
+  if (!searchString) return [];
+
+  // We approximate case-insensitive search by converting both needles and
+  // haystacks to lowercase.
+  const needle = searchString.toLowerCase();
+
+  // Augment users with whether assorted fields match the search string or not, and where.
+  const augmentedUsers: AugmentedUser[] = users.map((u) => {
```

Comment:
If I'm not mistaken, we currently only subscribe to other users' display names, so these other filters can only ever return yourself.

Reply:
IIRC, https://github.com/deathandmayhem/jolly-roger/blob/main/imports/server/users.ts#L12-L28 ensures that you always have your own profile information available (and I don't think any of our subscriptions explicitly _exclude_ self anyway)

Reply:
Sorry, my commend was ambiguously worded. It definitely searches your own profile. The problem (I think...) is that it can't search email addresses (or other fields) of other users because it doesn't subscribe to that information, so the results are display name matches for anyone and any kind of match for yourself.

Reply:
Hmm. Yeah, I think that's right. It effectively means we need a subscription to `huntProfiles`.

Reply:
Ah, good catch.  I suppose this would be the sort of thing that we could have the editor itself subscribe to, since nothing else on the page cares about the non-displayName fields for other users.

Alternately, we could say that completions have to target `displayName`?

Reply:
I think right now this also lets you mention someone not subscribed to this hunt, which is at best weird and at worst an easy way to accidentally send spoilers.

Reply:
Yeah, although that's an easy fix - I'll post a suggestion.

Oh - for what it's worth, we don't just subscribe to `displayNames`. `CallSection` subscribes to `avatars`, which means we have discord account info, so that's already available, though sort of by accident. I think it'd be pretty reasonable to subscribe to `huntProfiles` and have all the info we want, although I would be happiest if we could make it so that the puzzle page _only_ subscribed to `huntProfiles` and didn't _also_ subscribe to `displayNames` and `avatars`

Reply:
Hmmm.  `huntProfiles` is a fair bit more data than `displayNames`, and it's used only by the chat input, so I feel a little bad having the other components on the page block on "all the data of all the people" when rendering the chat log and the rest of the page could proceed with `huntProfiles` getting loaded lazily in the background without triggering rerender thrashing for the whole PuzzlePage.

I think my current approach will be to sub to `huntProfiles` in `ChatInput`, far down in the component hierarchy and to leave the `displayNames` and `avatars` subs where they are for now.